File: objectModel/CSharp/Microsoft.CommonDataModel.ObjectModel/Cdm/Projections/CdmOperationAddArtifactAttribute.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

namespace Microsoft.CommonDataModel.ObjectModel.Cdm
{
    using Microsoft.CommonDataModel.ObjectModel.Enums;
    using Microsoft.CommonDataModel.ObjectModel.ResolvedModel;
    using Microsoft.CommonDataModel.ObjectModel.Utilities;
    using Microsoft.CommonDataModel.ObjectModel.Utilities.Logging;
    using System;
    using System.Collections.Generic;
    using System.Linq;

    /// <summary>
    /// Class to handle AddArtifactAttribute operations
    /// </summary>
    public class CdmOperationAddArtifactAttribute : CdmOperationBase
    {
        private static readonly string Tag = nameof(CdmOperationAddArtifactAttribute);

        public CdmAttributeItem NewAttribute { get; set; }

        public bool? InsertAtTop { get; set; }

        public CdmOperationAddArtifactAttribute(CdmCorpusContext ctx) : base(ctx)
        {
            this.ObjectType = CdmObjectType.OperationAddArtifactAttributeDef;
            this.Type = CdmOperationType.AddArtifactAttribute;
        }

        /// <inheritdoc />
        public override CdmObject Copy(ResolveOptions resOpt = null, CdmObject host = null)
        {
            if (resOpt == null)
            {
                resOpt = new ResolveOptions(this, this.Ctx.Corpus.DefaultResolutionDirectives);
            }

            var copy = host == null ? new CdmOperationAddArtifactAttribute(this.Ctx) : host as CdmOperationAddArtifactAttribute;

            copy.NewAttribute = this.NewAttribute?.Copy(resOpt) as CdmAttributeItem;
            copy.InsertAtTop = this.InsertAtTop;

            this.CopyProj(resOpt, copy);
            return copy;
        }

        /// <inheritdoc />
        [Obsolete("CopyData is deprecated. Please use the Persistence Layer instead.")]
        public override dynamic CopyData(ResolveOptions resOpt = null, CopyOptions options = null)
        {
            return CdmObjectBase.CopyData<CdmOperationAddArtifactAttribute>(this, resOpt, options);
        }

        public override string GetName()
        {
            return "operationAddArtifactAttribute";
        }

        /// <inheritdoc />
        [Obsolete]
        public override CdmObjectType GetObjectType()
        {
            return CdmObjectType.OperationAddArtifactAttributeDef;
        }

        /// <inheritdoc />
        public override bool Validate()
        {
            List<string> missingFields = new List<string>();

            if (this.NewAttribute == null)
            {
                missingFields.Add(nameof(this.NewAttribute));
            }

            if (missingFields.Count > 0)
            {
                Logger.Error(this.Ctx, Tag, nameof(Validate), this.AtCorpusPath, CdmLogCode.ErrValdnIntegrityCheckFailure, this.AtCorpusPath, string.Join(", ", missingFields.Select((s) =>$"'{s}'")));
                return false;
            }

            return true;
        }

        /// <inheritdoc />
        public override bool Visit(string pathFrom, VisitCallback preChildren, VisitCallback postChildren)
        {
            string path = this.UpdateDeclaredPath(pathFrom);

            if (preChildren?.Invoke(this, path) == true)
            {
                return false;
            }

            if (this.NewAttribute != null && this.NewAttribute.Visit($"{path}/newAttribute/", preChildren, postChildren))
                return true;

            if (postChildren != null && postChildren.Invoke(this, path))
                return true;

            return false;
        }

        /// <inheritdoc />
        internal override ProjectionAttributeStateSet AppendProjectionAttributeState(
            ProjectionContext projCtx,
            ProjectionAttributeStateSet projOutputSet,
            CdmAttributeContext attrCtx)
        {
            if (this.InsertAtTop != true)
            {
                AddAllPreviousAttributeStates(projCtx, projOutputSet);
                AddNewArtifactAttributeState(projCtx, projOutputSet, attrCtx);
            } else
            {
                AddNewArtifactAttributeState(projCtx, projOutputSet, attrCtx);
                AddAllPreviousAttributeStates(projCtx, projOutputSet);
            }

            return projOutputSet;
        }

        /// <summary>
        /// Create a new artifact attribute and add it to the projOutputSet.
        /// </summary>
        /// <param name="projCtx">The projection context.</param>
        /// <param name="projOutputSet">The projection attribute state set.</param>
        /// <param name="attrCtx">The attribute context.</param>
        /// <returns></returns>
        private void AddNewArtifactAttributeState(ProjectionContext projCtx, ProjectionAttributeStateSet projOutputSet, CdmAttributeContext attrCtx)
        {
            // Create a new attribute context for the operation
            AttributeContextParameters attrCtxOpAddArtifactAttrParam = new AttributeContextParameters
            {
                under = attrCtx,
                type = CdmAttributeContextType.OperationAddArtifactAttribute,
                Name = $"operation/index{Index}/{this.GetName()}"
            };
            CdmAttributeContext attrCtxOpAddArtifactAttr = CdmAttributeContext.CreateChildUnder(projCtx.ProjectionDirective.ResOpt, attrCtxOpAddArtifactAttrParam);

            if (this.NewAttribute is CdmTypeAttributeDefinition)
            {
                // Create a new attribute context for the new artifact attribute we will create
                AttributeContextParameters attrCtxNewAttrParam = new AttributeContextParameters
                {
                    under = attrCtxOpAddArtifactAttr,
                    type = CdmAttributeContextType.AddedAttributeNewArtifact,
                    Name = this.NewAttribute.FetchObjectDefinitionName()
                };
                CdmAttributeContext attrCtxNewAttr = CdmAttributeContext.CreateChildUnder(projCtx.ProjectionDirective.ResOpt, attrCtxNewAttrParam);
                ResolvedAttribute newResAttr = CreateNewResolvedAttribute(projCtx, attrCtxNewAttr, (CdmAttribute)this.NewAttribute);

                // Create a new projection attribute state for the new artifact attribute and add it to the output set
                // There is no previous state for the newly created attribute
                ProjectionAttributeState newPAS = new ProjectionAttributeState(projOutputSet.Ctx)
                {
                    CurrentResolvedAttribute = newResAttr
                };

                projOutputSet.Add(newPAS);
            }
            else if (this.NewAttribute is CdmEntityAttributeDefinition || this.NewAttribute is CdmAttributeGroupReference)
            {
                var typeStr = this.NewAttribute is CdmEntityAttributeDefinition ? "an entity attribute" : "an attribute group";
                Logger.Warning(this.Ctx, Tag, nameof(AppendProjectionAttributeState), this.AtCorpusPath, CdmLogCode.WarnProjAddArtifactAttrNotSupported, typeStr);
            }
            else
            {
                Logger.Error(this.Ctx, Tag, nameof(AppendProjectionAttributeState), this.AtCorpusPath, CdmLogCode.ErrProjUnsupportedSource, this.NewAttribute.ObjectType.ToString(), this.GetName());
            }
        }

        /// <summary>
        /// Pass through all the input projection attribute states if there are any.
        /// </summary>
        /// <param name="projCtx">The projection context.</param>
        /// <param name="projOutputSet">The projection attribute state set.</param>
        /// <returns></returns>
        private void AddAllPreviousAttributeStates(ProjectionContext projCtx, ProjectionAttributeStateSet projOutputSet)
        {
            // Pass through all the input projection attribute states if there are any
            foreach (ProjectionAttributeState currentPAS in projCtx.CurrentAttributeStateSet.States)
            {
                projOutputSet.Add(currentPAS);
            }
        }
    }
}

File: objectModel/TypeScript/Cdm/Projections/CdmOperationAddArtifactAttribute.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

import {
    AttributeContextParameters,
    CdmAttributeContext,
    cdmAttributeContextType,
    CdmAttributeItem,
    CdmCorpusContext,
    cdmLogCode,
    CdmObject,
    cdmObjectType,
    CdmOperationBase,
    cdmOperationType,
    Logger,
    ProjectionAttributeState,
    ProjectionAttributeStateSet,
    ProjectionContext,
    resolveOptions,
    VisitCallback
} from '../../internal';
import { CdmAttributeGroupReference } from '../CdmAttributeGroupReference';
import { CdmEntityAttributeDefinition } from '../CdmEntityAttributeDefinition';
import { CdmTypeAttributeDefinition } from '../CdmTypeAttributeDefinition';

/**
 * Class to handle AddArtifactAttribute operations
 */
export class CdmOperationAddArtifactAttribute extends CdmOperationBase {
    private TAG: string = CdmOperationAddArtifactAttribute.name;

    public newAttribute: CdmAttributeItem;
    public insertAtTop?: boolean;

    constructor(ctx: CdmCorpusContext) {
        super(ctx);
        this.objectType = cdmObjectType.operationAddArtifactAttributeDef;
        this.type = cdmOperationType.addArtifactAttribute;
    }

    /**
     * @inheritdoc
     */
    public copy(resOpt?: resolveOptions, host?: CdmObject): CdmObject {
        if (!resOpt) {
            resOpt = new resolveOptions(this, this.ctx.corpus.defaultResolutionDirectives);
        }

        const copy: CdmOperationAddArtifactAttribute = !host ? new CdmOperationAddArtifactAttribute(this.ctx) : host as CdmOperationAddArtifactAttribute;

        copy.newAttribute = this.newAttribute ? this.newAttribute.copy(resOpt) as CdmAttributeItem : undefined;
        copy.insertAtTop = this.insertAtTop;
        
        this.copyProj(resOpt, copy);
        return copy;
    }

    /**
     * @inheritdoc
     */
    public getName(): string {
        return 'operationAddArtifactAttribute';
    }

    /**
     * @inheritdoc
     */
    public getObjectType(): cdmObjectType {
        return cdmObjectType.operationAddArtifactAttributeDef;
    }

    /**
     * @inheritdoc
     */
    public validate(): boolean {
        const missingFields: string[] = [];

        if (!this.newAttribute) {
            missingFields.push('newAttribute');
        }

        if (missingFields.length > 0) {
            Logger.error(this.ctx, this.TAG, this.validate.name, this.atCorpusPath, cdmLogCode.ErrValdnIntegrityCheckFailure, this.atCorpusPath, missingFields.map((s: string) => `'${s}'`).join(', '));
            return false;
        }

        return true;
    }

    /**
     * @inheritdoc
     */
    public visit(pathFrom: string, preChildren: VisitCallback, postChildren: VisitCallback): boolean {
        const path = this.fetchDeclaredPath(pathFrom);

        if (preChildren && preChildren(this, path)) {
            return false;
        }

        if (this.newAttribute !== undefined && this.newAttribute.visit(`${path}/newAttribute/`, preChildren, postChildren)) {
            return true;
        }

        if (postChildren && postChildren(this, path)) {
            return true;
        }

        return false;
    }

    /**
     * @inheritdoc
     * @internal
     */
    public appendProjectionAttributeState(projCtx: ProjectionContext, projOutputSet: ProjectionAttributeStateSet, attrCtx: CdmAttributeContext): ProjectionAttributeStateSet {
        if (this.insertAtTop != true) {
            this.addAllPreviousAttributeStates(projCtx, projOutputSet);
            this.addNewArtifactAttributeState(projCtx, projOutputSet, attrCtx);
        } else {
            this.addNewArtifactAttributeState(projCtx, projOutputSet, attrCtx);
            this.addAllPreviousAttributeStates(projCtx, projOutputSet);
        }

        return projOutputSet;
    }
    
    private addNewArtifactAttributeState(projCtx: ProjectionContext, projOutputSet: ProjectionAttributeStateSet, attrCtx: CdmAttributeContext): ProjectionAttributeStateSet {
        // Create a new attribute context for the operation
        const attrCtxOpAddAttrGroupParam: AttributeContextParameters = {
            under: attrCtx,
            type: cdmAttributeContextType.operationAddArtifactAttribute,
            name: `operation/index${this.index}/${this.getName()}`
        };
        const attrCtxOpAddArtifactAttr: CdmAttributeContext = CdmAttributeContext.createChildUnder(projCtx.projectionDirective.resOpt, attrCtxOpAddAttrGroupParam);

        if (this.newAttribute instanceof CdmTypeAttributeDefinition) {
            // Create a new attribute context for the new artifact attribute we will create
            const attrCtxAttrGroupParam: AttributeContextParameters = {
                under: attrCtxOpAddArtifactAttr,
                type: cdmAttributeContextType.addedAttributeNewArtifact,
                name: this.newAttribute.fetchObjectDefinitionName()
            };
            const attrCtxNewAttr: CdmAttributeContext = CdmAttributeContext.createChildUnder(projCtx.projectionDirective.resOpt, attrCtxAttrGroupParam);
            const newResAttr = CdmOperationBase.createNewResolvedAttribute(projCtx, attrCtxNewAttr, this.newAttribute);
    
            // Create a new projection attribute state for the new artifact attribute and add it to the output set
            // There is no previous state for the newly created attribute
            const newPAS: ProjectionAttributeState = new ProjectionAttributeState(projOutputSet.ctx);
            newPAS.currentResolvedAttribute = newResAttr;
            
            projOutputSet.add(newPAS);
        } else if (this.newAttribute instanceof CdmEntityAttributeDefinition || this.newAttribute instanceof CdmAttributeGroupReference) {
            const typeStr: string = this.newAttribute instanceof CdmEntityAttributeDefinition ? 'an entity attribute' : 'an attribute group';
            Logger.warning(this.ctx, this.TAG, this.appendProjectionAttributeState.name, this.atCorpusPath, cdmLogCode.WarnProjAddArtifactAttrNotSupported, typeStr);
        } else {
            Logger.warning(this.ctx, this.TAG, this.appendProjectionAttributeState.name, this.atCorpusPath, cdmLogCode.ErrProjUnsupportedSource, this.newAttribute.objectType.toString(), this.getName());
        }


        return projOutputSet;
    }

    /**
     * Pass through all the input projection attribute states if there are any.
     */
    private addAllPreviousAttributeStates(projCtx: ProjectionContext, projOutputSet: ProjectionAttributeStateSet): void {
        // Pass through all the input projection attribute states if there are any
        for (const currentPAS of projCtx.currentAttributeStateSet.states) {
            projOutputSet.add(currentPAS);
        }
    }
}
